Mark message invalid when it fails enablement check or it is not handled

szewai opened the pull request:
#### ac6dcc85f22e9190982ed5f7a86f6383d113dcdc
<pre>
Mark message invalid when it fails enablement check or it is not handled
<a href="https://bugs.webkit.org/show_bug.cgi?id=283394">https://bugs.webkit.org/show_bug.cgi?id=283394</a>

Reviewed by Ryosuke Niwa.

The patch makes a few changes to make generated message receiver code more consistent:
1. Invoke Decoder::markInvalid(), instead of Connection::markCurrentlyDispatchedMessageAsInvalid() to mark message as
invalid. This enables us to mark a message dispatched to stream connection invalid.
2. Remove the checks for ignoreInvalidMessageForTesting, as Connection already checks that before notifying client about
invalid message (see Connection::dispatchMessage and StreamServerConnection::dispatchStreamMessage).
3. Convert ASSERT_NOT_REACHED_WITH_MESSAGE to RELEASE_LOG_ERROR and markInvalid(), as it makes little sense to crash
receiver process when it receives an invalid message (including message that cannot be handled). For invalid message,
the receiver proess may request the sender process to be killed, or just ignore the message if
ignoreInvalidMessageForTesting is true.

* Source/WebKit/Scripts/webkit/messages.py:
(async_message_statement):
(generate_enabled_by_for_receiver):
(generate_message_handler):
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessageReceiver.cpp:
(WebKit::TestWithCVPixelBuffer::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByAndConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByOrConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
(WebKit::TestWithEnabledIf::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessageReceiver.cpp:
(WebKit::TestWithIfMessage::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessageReceiver.cpp:
(WebKit::TestWithImageData::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
(WebKit::TestWithLegacyReceiver::didReceiveMessage):
(WebKit::TestWithLegacyReceiver::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessageReceiver.cpp:
(WebKit::TestWithSemaphore::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessageReceiver.cpp:
(WebKit::TestWithStreamBatched::didReceiveStreamMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp:
(WebKit::TestWithStreamBuffer::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessageReceiver.cpp:
(WebKit::TestWithStream::didReceiveStreamMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamServerConnectionHandleMessageReceiver.cpp:
(WebKit::TestWithStreamServerConnectionHandle::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithWantsAsyncDispatchMessageReceiver.cpp:
(WebKit::TestWithWantsAsyncDispatch::didReceiveMessage):
(WebKit::TestWithWantsAsyncDispatch::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchMessageReceiver.cpp:
(WebKit::TestWithWantsDispatch::didReceiveMessage):
(WebKit::TestWithWantsDispatch::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchNoSyncMessagesMessageReceiver.cpp:
(WebKit::TestWithWantsDispatchNoSyncMessages::didReceiveMessage):
(WebKit::TestWithWantsDispatchNoSyncMessages::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:
(WebKit::TestWithoutAttributes::didReceiveMessage):
(WebKit::TestWithoutAttributes::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessageReceiver.cpp:
(WebKit::TestWithoutUsingIPCConnection::didReceiveMessageWithReplyHandler):
* Source/WebKit/Scripts/webkit/tests/WebKitPlatformGeneratedSerializers.cpp:
(IPC::NSSomeOtherFoundationType&gt;::encode):

Canonical link: <a href="https://commits.webkit.org/286885@main">https://commits.webkit.org/286885@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d06bc8c5c36d0aab8de894d9aeb5f4f0bf783c8a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56398 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81972 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28645 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65546 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4695 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60626 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18641 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80430 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50591 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66408 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40924 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/76869 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47993 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26968 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24244 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83352 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4743 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3225 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68879 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66378 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68137 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17022 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12133 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10236 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4690 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7505 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4709 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8144 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
```diff
@@ -50,27 +48,21 @@ void TestWithEnabledBy::didReceiveMessage(IPC::Connection& connection, IPC::Deco
         if (sharedPreferences && sharedPreferences->someFeature)
             return IPC::handleMessage<Messages::TestWithEnabledBy::ConditionallyEnabled>(connection, decoder, this, &TestWithEnabledBy::conditionallyEnabled);
-        runtimeEnablementCheckFailed = true;
+        return decoder.markInvalid();
```
Don't we want to add a release log here?

szewai replied:
Will add.

rniwa commented on Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
```diff
@@ -50,27 +48,21 @@ void TestWithEnabledBy::didReceiveMessage(IPC::Connection& connection, IPC::Deco
         if (sharedPreferences && (sharedPreferences->someFeature && sharedPreferences->otherFeature))
             return IPC::handleMessage<Messages::TestWithEnabledBy::ConditionallyEnabledAnd>(connection, decoder, this, &TestWithEnabledBy::conditionallyEnabledAnd);
-        runtimeEnablementCheckFailed = true;
+        return decoder.markInvalid();
```
ditto.

rniwa commented on Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
```diff
@@ -50,27 +48,21 @@ void TestWithEnabledBy::didReceiveMessage(IPC::Connection& connection, IPC::Deco
         if (sharedPreferences && (sharedPreferences->someFeature || sharedPreferences->otherFeature))
             return IPC::handleMessage<Messages::TestWithEnabledBy::ConditionallyEnabledOr>(connection, decoder, this, &TestWithEnabledBy::conditionallyEnabledOr);
-        runtimeEnablementCheckFailed = true;
+        return decoder.markInvalid();
```
ditto.

rniwa commented on Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
```diff
@@ -59,17 +53,11 @@ void TestWithEnabledIf::didReceiveMessage(IPC::Connection& connection, IPC::Deco
         if (featureEnabled())
             return IPC::handleMessage<Messages::TestWithEnabledIf::OnlyEnabledIfFeatureEnabled>(connection, decoder, this, &TestWithEnabledIf::onlyEnabledIfFeatureEnabled);
-        runtimeEnablementCheckFailed = true;
+        return decoder.markInvalid();
```
release log?

szewai committed:
Mark message invalid when it fails enablement check or it is not handled
https://bugs.webkit.org/show_bug.cgi?id=283394

Reviewed by Ryosuke Niwa.

The patch makes a few changes to make generated message receiver code more consistent:
1. Invoke Decoder::markInvalid(), instead of Connection::markCurrentlyDispatchedMessageAsInvalid() to mark message as
invalid. This enables us to mark a message dispatched to stream connection invalid.
2. Remove the checks for ignoreInvalidMessageForTesting, as Connection already checks that before notifying client about
invalid message (see Connection::dispatchMessage and StreamServerConnection::dispatchStreamMessage).
3. Convert ASSERT_NOT_REACHED_WITH_MESSAGE to RELEASE_LOG_ERROR and markInvalid(), as it makes little sense to crash
receiver process when it receives an invalid message (including message that cannot be handled). For invalid message,
the receiver proess may request the sender process to be killed, or just ignore the message if
ignoreInvalidMessageForTesting is true.

* Source/WebKit/Scripts/webkit/messages.py:
(async_message_statement):
(generate_enabled_by_for_receiver):
(generate_message_handler):
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessageReceiver.cpp:
(WebKit::TestWithCVPixelBuffer::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByAndConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
(WebKit::TestWithEnabledBy::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessageReceiver.cpp:
(WebKit::TestWithEnabledByOrConjunction::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessageReceiver.cpp:
(WebKit::TestWithEnabledIf::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessageReceiver.cpp:
(WebKit::TestWithIfMessage::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessageReceiver.cpp:
(WebKit::TestWithImageData::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
(WebKit::TestWithLegacyReceiver::didReceiveMessage):
(WebKit::TestWithLegacyReceiver::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessageReceiver.cpp:
(WebKit::TestWithSemaphore::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessageReceiver.cpp:
(WebKit::TestWithStreamBatched::didReceiveStreamMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp:
(WebKit::TestWithStreamBuffer::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessageReceiver.cpp:
(WebKit::TestWithStream::didReceiveStreamMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamServerConnectionHandleMessageReceiver.cpp:
(WebKit::TestWithStreamServerConnectionHandle::didReceiveMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithWantsAsyncDispatchMessageReceiver.cpp:
(WebKit::TestWithWantsAsyncDispatch::didReceiveMessage):
(WebKit::TestWithWantsAsyncDispatch::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchMessageReceiver.cpp:
(WebKit::TestWithWantsDispatch::didReceiveMessage):
(WebKit::TestWithWantsDispatch::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchNoSyncMessagesMessageReceiver.cpp:
(WebKit::TestWithWantsDispatchNoSyncMessages::didReceiveMessage):
(WebKit::TestWithWantsDispatchNoSyncMessages::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:
(WebKit::TestWithoutAttributes::didReceiveMessage):
(WebKit::TestWithoutAttributes::didReceiveSyncMessage):
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessageReceiver.cpp:
(WebKit::TestWithoutUsingIPCConnection::didReceiveMessageWithReplyHandler):
* Source/WebKit/Scripts/webkit/tests/WebKitPlatformGeneratedSerializers.cpp:
(IPC::NSSomeOtherFoundationType>::encode):

Canonical link: https://commits.webkit.org/286885@main
